lib/src/ioctl/g_ext_ctrls.rs

```diff
@@ -0,0 +1,418 @@
+    }
+}
+
+/// Get a single extended control
+///
+/// While vidioc_g_ext_ctrls() accepts an array of controls,
+/// the kernel internally seems not to commit all the values atomically,
+/// and proceeds only until the first failure.
+///
+/// Given the above we provide an interface for querying a single control,
+/// as this greatly simplifies the code.
+pub fn g_ext_ctrl<F: AsRawFd, R: AsRawFd>(
+    fd: &F,
+    request_fd: Option<&R>,
+    ctrl_kind: ExtControlKind,
+) -> Result<ExtControl, ExtControlError> {
+    let mut control = bindings::v4l2_ext_control {
+        // SAFETY: ok to zero-fill this struct, the pointer it contains will be assigned to in this function
+        ..unsafe { mem::zeroed() }
+    };
+    let mut controls = bindings::v4l2_ext_controls {
+        // SAFETY: ok to zero-fill this struct, the pointer it contains will be assigned to in this function
+        ..unsafe { mem::zeroed() }
+    };
+
+    let mut compound_ctrl = CompoundControl::try_from(ctrl_kind)?;
+
+    control.id = ctrl_kind as u32;
+    control.size = compound_ctrl.size() as u32;
+    // the pointer is assigned a proper value
+    control.__bindgen_anon_1.ptr = compound_ctrl.as_mut_ptr();
+
+    controls.__bindgen_anon_1.which = bindings::V4L2_CTRL_WHICH_CUR_VAL;
+    controls.count = 1;
+    // the pointer is assigned a proper value
+    controls.controls = &mut control;
+    if let Some(request_fd) = request_fd {
+        controls.request_fd = request_fd.as_raw_fd();
+    }
+
+    // SAFETY: the 'controls' argument is properly set up above
+    match unsafe { ioctl::vidioc_g_ext_ctrls(fd.as_raw_fd(), &mut controls) } {
+        Ok(_) => Ok(ExtControl::try_from(compound_ctrl)?),
+        Err(e) => Err(ExtControlError::IoctlError(e)),
+    }
+}
+
+/// Set a single extended control
+///
+/// While vidioc_s_ext_ctrls() accepts an array of controls,
+/// the kernel internally seems not to commit all the values atomically,
+/// and proceeds only until the first failure.
+///
+/// Given the above we provide an interface for setting a single control,
+/// as this greatly simplifies the code.
+pub fn s_ext_ctrl<F: AsRawFd, R: AsRawFd>(
+    fd: &F,
+    request_fd: Option<&R>,
+    ctrl: &ExtControl,
+) -> Result<ExtControl, ExtControlError> {
+    let mut control = bindings::v4l2_ext_control {
+        // SAFETY: ok to zero-fill this struct, the pointer it contains will be assigned to in this function
+        ..unsafe { mem::zeroed() }
+    };
+    let mut controls = bindings::v4l2_ext_controls {
+        // SAFETY: ok to fill this struct, the pointer it contains will be assigned to in this function
+        ..unsafe { mem::zeroed() }
+    };
+
+    let mut compound_ctrl = CompoundControl::try_from(ctrl)?;
+
+    control.id = ExtControlKind::from(ctrl) as u32;
+    control.size = compound_ctrl.size() as u32;
+    // the pointer is assigned a proper value
+    control.__bindgen_anon_1.ptr = compound_ctrl.as_mut_ptr();
```

Comment:
This is potentially dangerous, as you have no guarantee about the data layout of `compound_ctrl` until it is tagged with `#[repr(C)]`. 

I'm trying to think about a way to get rid of `CompoundControl` altogether, as it plays a duplicate role with `ExtControl` somehow, let's see if that can solve this issue as well.

Reply:
The problem we've been trying to solve here is having to allocate a varible amount of space for different kinds of C structs. Using CompoundControl we avoided Box'ed allocations. But if you can come up with yet another (and better) idea, then let's change it.